src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/FhirPathPatch/FhirPathPatchBuilder.cs

```diff
@@ -0,0 +1,155 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using FhirPathPatch.Operations;
+using Hl7.Fhir.Model;
+using static Hl7.Fhir.Model.Parameters;
+
+namespace FhirPathPatch
+{
+    /// <summary>
+    /// Handles patching a FHIR Resource in a builder pattern manner.
+    /// </summary>
+    public class FhirPathPatchBuilder
+    {
+        private Resource resource;
+
+        private List<PendingOperation> operations;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="FhirPathPatchBuilder"/> class.
+        /// </summary>
+        /// <param name="resource">FHIR Resource.</param>
+        public FhirPathPatchBuilder(Resource resource)
+        {
+            this.resource = resource;
+
+            this.operations = new List<PendingOperation>();
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="FhirPathPatchBuilder"/> class with Patch Parameters.
+        /// </summary>
+        /// <param name="resource">FHIR Resource.</param>
+        /// <param name="parameters">Patch Parameters.</param>
+        public FhirPathPatchBuilder(Resource resource, Parameters parameters)
+        : this(resource)
+        {
+            this.Build(parameters);
+        }
+
+        /// <summary>
+        /// Applies the list of pending operations to the resource.
+        /// </summary>
+        public Resource Apply()
```

Comment:
>Apply [](http://example.com/codeflow?start=24&length=5)

Since we use Firely maybe it make sense to check out their patch service implementation: [https://github.com/FirelyTeam/spark/blob/r4/master/src/Spark.Engine/Service/FhirServiceExtensions/PatchService.cs](https://github.com/FirelyTeam/spark/blob/r4/master/src/Spark.Engine/Service/FhirServiceExtensions/PatchService.cs)

Reply:
My intent in comment was to make sure we don't reinvent bike, and it's good to know what you check other implementation.

Reply:
@Ivanidzo4ka - did some digging. Spark's implementation doesn't pass any tests involving [Anonymous Types](https://github.com/FHIR/fhir-test-cases/blob/752b01313ecbc1e13a942e1b3e25c96b3f7f3449/r5/patch/fhir-path-tests.xml#L450) or [Nested Primitives](https://github.com/FHIR/fhir-test-cases/blob/752b01313ecbc1e13a942e1b3e25c96b3f7f3449/r5/patch/fhir-path-tests.xml#L297). I remember trying to fix Spark's PatchService code, but failing since the code is complex. I'm going to do a simple performance comparison later just to make sure the two approaches aren't wildly different. As of right now, the approach in this PR is superior because it meets all of the FHIRPatch tests defined by HL7.

Reply:
@Ivanidzo4ka - performance difference is negligible testing with 100 sequential FHIRPatch `add` operations and 100 sequential FHIRPatch `replace` operations.